### ISSUE TITLE
Concurrent access to multiple GPUs throws: An item with the same key has already been added

### ISSUE THREAD

**Issue by miyconst** [Sergio0694/ComputeSharp#63]
Since I have figured out that it's not possible to concurrently use the same GPU (see [Concurrent GPU access](https://github.com/Sergio0694/ComputeSharp/issues/62)), I am trying to at least use all available GPUs at the same time.

This results in the following exception:

```
An item with the same key has already been added. Key: (39086322, 32, 1, 1)
```

The exception is not always reproducible, thus there is a race condition.

Simple example to reproduce the issue:

```cs
public void UseAllGpusAtTheSameTime()
{
    int threads = 12;
    List<GraphicsDevice> gpus = Gpu.EnumerateDevices().ToList();
    List<Task> tasks = new List<Task>();
    Stopwatch stopwatch = new Stopwatch();

    stopwatch.Start();
    for (int i = 0; i < threads; i++)
    {
        GraphicsDevice gpu = gpus[i % gpus.Count];

        tasks.Add(Task.Run(() =>
        {
            Console.WriteLine($"GPU {gpu.Description.Description} started");

            int sum = 0;
            int count = 1000;

            lock (gpu)
            {
                ReadWriteBuffer<int> buffer = gpu.AllocateReadWriteBuffer<int>(count);

                gpu.For(count, thread =>
                {
                    buffer[thread.X] = thread.X;
                });

                sum = buffer.GetData().Sum();
            }

            Console.WriteLine($"GPU {gpu.Description.Description} finished, sum: {sum}");
        }));
    }
    Task.WhenAll(tasks).GetAwaiter().GetResult();
    stopwatch.Stop();
    Console.WriteLine($"Total elapsed: {stopwatch.Elapsed}");
}
```

In my setup I have two GPUs: GTX 1080 and GTX 1050 TI.

```
  Message: 
    System.ArgumentException : An item with the same key has already been added. Key: (23514884, 32, 1, 1)
  Stack Trace: 
    Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
    ShaderRunner.Run(GraphicsDevice device, Int32 x, Int32 y, Int32 z, Int32 threadsX, Int32 threadsY, Int32 threadsZ, Action`1 action)
    ShaderRunner.Run(GraphicsDevice device, Int32 x, Int32 y, Int32 z, Action`1 action)
    GraphicsDeviceExtensions.For(GraphicsDevice device, Int32 x, Action`1 action)
```

**Comment by miyconst** [comment]
The issue was present in `1.2.0`, but seems to be fixed in `1.3.0`. Closing for now, will reopen if I will be able to reproduce it with `1.3.0`.

**Comment by Sergio0694** [comment]
Hey, sorry for the delay! Yeah I did fix some concurrency issues with `1.3.0`, glad to hear this issue in particular has been fixed too! I'll follow up with you on your other issue, thank you for your help!
